### PR TITLE
Version 1 of a fully functioning search API

### PR DESCRIPTION
**Pull request by nshumoogum**
### What

See title

### How to review

Check unit tests pass and run locally by following README.d file, set configuration to point to develop environment to run against search service instant.

### CODE REVIEW

File: CourseSearchAPIHttpTrigger/search.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+        self.index_name = index_name
+
+    def get(self):
+        try:
+            url = self.url + "/indexes/" + self.index_name + "/docs" + self.query_string
+
+            response = requests.get(url, headers=self.headers)
+            response_body = response.json()
+
+            postcode_object = {
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/search.py]: Maybe a better name would be postcode_lat_long would be a better name?

File: CourseSearchAPIHttpTrigger/search.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    def __init__(self, url, api_key, api_version, index_name, postcode):
+        self.url = url
+        self.headers = {
+            "Content-Type": "application/json",
+            "api-key": api_key,
+            "odata": "verbose",
+        }
+
+        # lowercase postcode and remove whitespace
+        self.p = postcode.lower().replace(" ", "")
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/search.py]: Maybe self.postcode is a better name?

File: CourseSearchAPIHttpTrigger/search.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+PARENTDIR = os.path.dirname(CURRENTDIR)
+sys.path.insert(0, CURRENTDIR)
+sys.path.insert(0, PARENTDIR)
+
+import exceptions
+
+
+def find_postcode(url, api_key, api_version, index_name, postcode):
+
+    try:
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/search.py]: Not sure what the try except is doing here?

File: CourseSearchAPIHttpTrigger/search.py [hunk 4]
```diff
@@ -0,0 +1,116 @@
+import logging
+import requests
+import os
+import json
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/search.py]: Unused import

File: CourseSearchAPIHttpTrigger/query.py [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+
+        if "honours_award" in self.query_params:
+            if self.query_params["honours_award"]:
+                filters.append("course/honours_award_provision/code eq '1'")
+            else:
+                filters.append("course/honours_award_provision/code eq '0'")
+
+        if (
+            "length_of_courses" in self.query_params
+            and len(self.query_params["length_of_courses"]) > 0
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/query.py]: No need to used len here as an empty_string, list, or dict is False

File: CourseSearchAPIHttpTrigger/query.py [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+            search.append(self.course)
+
+        if len(search) < 1:
+            query += "&search=*"
+        else:
+            query += "&search=" + " + ".join(search)
+
+        # Create filter part of query
+        filters = list()
+        if "countries" in self.query_params and len(self.query_params["countries"]) > 0:
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/query.py]: Perhaps simplify as:

`...and self.query_params["countries"]`

File: CourseSearchAPIHttpTrigger/query.py [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+        query = ""
+        # Create search part of query
+        search = list()
+        if self.institution != "":
+            search.append(self.institution)
+
+        if self.course != "":
+            search.append(self.course)
+
+        if len(search) < 1:
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/query.py]: Maybe:

if search:
    query += "&search=" + " + ".join(search)
else:
    query += "&search=*"

File: CourseSearchAPIHttpTrigger/query.py [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+        self.query_params = query_params
+
+    def build(self):
+        query = ""
+        # Create search part of query
+        search = list()
+        if self.institution != "":
+            search.append(self.institution)
+
+        if self.course != "":
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/query.py]: As above

File: CourseSearchAPIHttpTrigger/query.py [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+        self.course = course
+        self.institution = institution
+        self.institutions = institutions
+        self.query_params = query_params
+
+    def build(self):
+        query = ""
+        # Create search part of query
+        search = list()
+        if self.institution != "":
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/query.py]: Could be written more simply as:

if self.institution:
    search.append(self.institution)

File: CourseSearchAPIHttpTrigger/query.py [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+        return query.add_paging()
+    except Exception:
+        raise
+
+
+def build_institution_search_query(
+    course, institution, institutions, postcode_object, query_params
+):
+
+    try:
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/query.py]: as above.

File: CourseSearchAPIHttpTrigger/query.py [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+        query = Query(course, institution, institutions, postcode_object, query_params)
+
+        query.build()
+
+        return query.add_paging()
+    except Exception:
+        raise
+
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/query.py:14]: Not sure what the try except is doing here?

File: CourseSearchAPIHttpTrigger/models/error.py [hunk 12]
```diff
@@ -0,0 +1,37 @@
+import json
+
+err_limit_wrong_type = "limit value needs to be a number"
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/models/error.py]: Not an issue, but normal convention is UPPER_CASE naming style for constants

File: CourseSearchAPIHttpTrigger/helper.py [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+    total_courses = 0
+    total_institutions = 0
+
+    lower_range = requested_offset
+    upper_range = requested_limit + requested_offset
+
+    for facet in facets:
+        if total_institutions < lower_range:
+            offset += facet["count"]
+        elif total_institutions >= lower_range and total_institutions < upper_range:
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/helper.py]: or 

`elif lower_range <= total_institutions < upper_range`

File: CourseSearchAPIHttpTrigger/__init__.py [hunk 14]
```diff
@@ -1,24 +1,161 @@
...
+        )
+
+        facets = response_with_facets.json()
+
+        counts = {}
+        # Step 6 - handle facets to build correct
+        # limit and offset for next query
+        query_params["limit"], query_params["offset"], counts["institutions"], counts[
+            "courses"
+        ] = helper.get_offset_and_limit(
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/__init__.py:119]: Seems though get_offset_and_limit does more than just get offset and limit. Consider splitting into separate functions or renaming perhaps?

File: CourseSearchAPIHttpTrigger/__init__.py [hunk 15]
```diff
@@ -1,24 +1,161 @@
...
+
+            postcode_object["distance"] = convert_to_km(p[1])
+
+        # Step 2 - Validate query parameters
+        query_params, error_objects = validation.check_query_parameters(
+            countries, filters, length_of_course, limit, max_default_limit, offset
+        )
+
+        logging.info(f"query_params: {query_params}")
+        if len(error_objects) > 0:
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/__init__.py]: Might look better as:
    `if error_objects:`

File: CourseSearchAPIHttpTrigger/__init__.py [hunk 16]
```diff
@@ -1,24 +1,161 @@
...
+        filters = req.params.get("filters", "")
+        postcode = req.params.get("postcode", "")
+        institutions = req.params.get("institutions", "")
+        countries = req.params.get("countries", "")
+        length_of_course = req.params.get("length_of_course", "")
+
+        postcode_object = {}
+        # Step 1 Lookup postcode if parameter is set
+        if postcode != "":
+            p = postcode.split(",")
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/__init__.py]: Perhaps a more descriptive variable name would be helpful here?

File: CourseSearchAPIHttpTrigger/__init__.py [hunk 17]
```diff
@@ -1,24 +1,161 @@
 import logging
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/__init__.py:1]: Not used

**Review by philratcliffe — CHANGES_REQUESTED** [review]

File: CourseSearchAPIHttpTrigger/search.py [hunk 18]
```diff
@@ -0,0 +1,116 @@
...
+            "odata": "verbose",
+        }
+
+        # lowercase postcode and remove whitespace
+        self.p = postcode.lower().replace(" ", "")
+        self.query_string = "?api-version=" + api_version + "&search=" + self.p
+        self.index_name = index_name
+
+    def get(self):
+        try:
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/search.py:39]: maybe get_lat_long_for_postcode?

File: CourseSearchAPIHttpTrigger/__init__.py [hunk 19]
```diff
@@ -1,24 +1,161 @@
...
+            )
+
+    except Exception as e:
+        logging.error(traceback.format_exc())
+
+        # Raise so Azure sends back the HTTP 500
+        raise e
+
+
+def convert_to_km(distance):
```
Comment (philratcliffe) [inline CourseSearchAPIHttpTrigger/__init__.py]: convert_miles_to_km and distance_in_miles would be better?

File: CourseSearchAPIHttpTrigger/search.py (hunk 3 above)
Comment (nshumoogum) [reply CourseSearchAPIHttpTrigger/search.py]: Removing after conversation with @philratcliffe - no need to pass exceptions up the chain; this happens by default

File: CourseSearchAPIHttpTrigger/query.py (hunk 7 above)
Comment (nshumoogum) [reply CourseSearchAPIHttpTrigger/query.py]: I assume you mean more like this:
```
    search = list()
        if self.institution:
            search.append(self.institution)

        if self.course:
            search.append(self.course)

        if search:
            query += "&search=" + " + ".join(search)
        else:
            query += "&search=*"
```

File: CourseSearchAPIHttpTrigger/query.py (hunk 8 above)
Comment (nshumoogum) [reply CourseSearchAPIHttpTrigger/query.py]: See above

File: CourseSearchAPIHttpTrigger/__init__.py (hunk 17 above)
Comment (nshumoogum) [reply CourseSearchAPIHttpTrigger/__init__.py:1]: Which import?

**Review by philratcliffe — APPROVED** [review]

**Pull request merged**